pkg/store/bucket.go

```diff
@@ -1641,6 +1645,55 @@ func (s *BucketStore) LabelNames(ctx context.Context, req *storepb.LabelNamesReq
 	}, nil
 }
 
+func (s *BucketStore) UpdateLabelNamesBloom(ctx context.Context) error {
+	g, _ := errgroup.WithContext(ctx)
+
+	var mtx sync.Mutex
+	names := make(map[string]struct{})
+
+	for _, b := range s.blocks {
+		b := b
+		indexr := b.indexReader()
```

Comment:
Do we know how expensive this call could be? if we're doing this every 10 seconds for every block indexed by a store-gw, it should be in-memory for the most part, correct?

Reply:
We will probably need production usage data to know the impact.

Reply:
Hmmm, it would interesting to see the cost. My understanding is that this issue occurs only in certain configurations. Often those configuration are known in advance. In the same time you might know it won't occur or it occurs only for certain labels and certain storage.

Also, is there any eventual consistency here? Where labels changed but we didn't know about it yet? What are the potential causes? I guess wrong accuracy?

At some point we have to evaluate if this complexity, extra bucket calls and network use is better than reverting the sorting overhaul we did in the beginning of this year.

Reply:
That's a good point, I didn't realize that the index reader will reach out to object storage. Should we read labels in parallel then? Possibly with some limit on to control memory usage.

Reply:
Can we read label names from the postings offset table? I think we should have everything already on disk with binary index headers: https://github.com/prometheus/prometheus/blob/main/tsdb/docs/format/index.md#postings-offset-table.

Reply:
You're right, we can get this information from the index header. Furthermore, we could even build these filters at block load time instead of polling every 5 seconds.